pysnn/learning.py

```diff
@@ -8,58 +11,59 @@
 #########################################################
 # Learning rule base class
 #########################################################
-class LearningRule(nn.Module):
-    r"""Base class for correlation based learning rules in spiking neural networks."""
-
-    def __init__(self, connection, lr):
-        super(LearningRule, self).__init__()
-        self.connection = connection
-        self.register_buffer("lr", torch.tensor(lr, dtype=torch.float))
-
-    def no_grad(self):
-        _set_no_grad(self)
-
-    def reset_state(self):
-        pass
-
-    def init_rule(self):
-        self.no_grad()
-        self.reset_state()
+class LearningRule:
+    r"""Base class for correlation based learning rules in spiking neural networks.
+    
+    Arguments:
+        layers (iterable): an iterable or :class:`dict` of :class:`dict`s. 
+            The latter is a dict that contains a :class:`pysnn.Connection`s state dict, a pre-synaptic :class:`pysnn.Neuron`s state dict, 
+            and a post-synaptic :class:`pysnn.Neuron`s state dict that together form a single layer. These objects their state's will be 
+            used for optimizing weights.
+            During initialization of a learning rule that inherits from this class it is supposed to select only the paramters it needs
+            from these objects.
+            The higher lever iterable or :class:`dict` contain groups that use the same parameter during training. This is analogous to
+            PyTorch optimizers' parameter groups.
+        defaults: A dict containing default hyper paramters. This is a placeholder for possible changes later on, these groups would work
+            exactly the same as those for PyTorch optimizers.
+    """
 
+    def __init__(self, layers, defaults):
+        self.defaults = defaults
+        self.layers = layers
 
-#########################################################
-# Fede STDP
-#########################################################
-class FedeSTDP(LearningRule):
-    r"""STDP version for Paredes Valles, performs mean operation over the batch 
-    dimension before weight update."""
+    def update_state(self):
+        r"""Update state parameters of LearningRule based on latest network forward pass."""
+        raise NotImplementedError
 
-    def __init__(self, connection, lr, w_init, a):
-        super(FedeSTDP, self).__init__(connection, lr)
-        self.register_buffer("w_init", torch.tensor(w_init, dtype=torch.float))
-        self.register_buffer("a", torch.tensor(a, dtype=torch.float))
+    def step(self):
```

Comment:
`step` to indicate that it's not actually "forward"?

Reply:
Two reasons, to differentiate it from a Module, but also since this is the naming convention in regular PyTorch.